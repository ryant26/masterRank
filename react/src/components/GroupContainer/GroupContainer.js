import React, {
    Component
} from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Model from '../../model/model';
import MemberCard from './MemberCard/TimerMemberCard';

class GroupContainer extends Component {

    constructor(props) {
        super(props);
    }

    componentDidMount() {
        Model.createNewGroup(this.props.preferredHeroes.heroes[0]);
    }

    componentWillUnmount() {
        Model.leaveGroup(this.props.group.groupId);
    }

    render() {
        return (
            <div className="GroupContainer">
             <div>==================================</div>
                {this.props.group.pending &&

                    this.props.group.pending.map((member, i) =>
                        <MemberCard
                            member={member}
                            number={(i + 1)}
                            pending={true}
                            key={[member.platformDisplayName, member.heroName]}
                        />
                    )
                }
             <div>==================================</div>
            </div>
        );
    }
};

GroupContainer.propTypes = {
    group: PropTypes.shape({
        groupId: PropTypes.number,
        members: PropTypes.arrayOf(PropTypes.shape({
            heroName: PropTypes.string,
            platformDisplayName: PropTypes.string,
            skillRating: PropTypes.number,
            stats: PropTypes.object,
        })),
        pending: PropTypes.array,
        leader: PropTypes.shape({
            heroName: PropTypes.string,
            platformDisplayName: PropTypes.string,
            skillRating: PropTypes.number,
            stats: PropTypes.object,
        }),
    }),
    preferredHeroes: PropTypes.shape({
        heroes: PropTypes.array
    }).isRequired,
    user: PropTypes.object.isRequired
};

const mapStateToProps = (state) => {
    return {
      group: state.group,
      preferredHeroes: state.preferredHeroes,
      user: state.user
    };
};

export default connect(mapStateToProps)(GroupContainer);