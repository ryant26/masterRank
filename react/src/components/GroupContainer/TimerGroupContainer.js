import React, {
    Component
} from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Model from '../../model/model';
import MemberCard from './MemberCard/MemberCard';

class TimerGroupContainer extends Component {

    constructor(props) {
        super(props);
        this.state = {
            timers: {},
        }

        this.cancelInvite = this.cancelInvite.bind(this);
        this.setTimers = this.setTimers.bind(this);
    }

    //TODO: CRUD for group
    componentDidMount() {
        Model.createNewGroup(this.props.preferredHeroes.heroes[0]);
    }

    componentWillUnmount() {
        Model.leaveGroup(this.props.group.groupId);
        for( let key in this.state.timers) {
            clearInterval(this.state.timers[key]);
        };
    }

    componentWillReceiveProps(nextProps) {
        this.setTimers(nextProps);
    }

    setTimers(nextProps) {
        if(nextProps.group.pending){
            let currentTimers = this.state.timers;
            //Assumes only one prop can update at a time
            nextProps.group.pending.map((member) => {
                let key = [member.platformDisplayName, member.heroName];
                if(!currentTimers[key])
                    //No timer exists
                    currentTimers[key] =
                        setInterval(() => {this.cancelInvite(member)}, 5000);
            });

            this.setState({
                timers: currentTimers
            })
        }
    }

    cancelInvite(member) {
        let key = [member.platformDisplayName, member.heroName];
        alert(`${key}`);
        Model.cancelInvite({
           platformDisplayName: member.platformDisplayName,
           heroName: member.heroName
        });
        clearInterval(this.state.timers[key]);
        delete this.state.timers[key];
    }

    render() {
        return (
            <div className="TimerGroupContainer">
             <div>==================================</div>
                {this.props.group.pending &&

                    this.props.group.pending.map((member, i) =>
                        <MemberCard member={member} number={(i + 1)} key={i}/>
                    )
                }
             <div>==================================</div>
            </div>
        );
    }
};

TimerGroupContainer.propTypes = {
    group: PropTypes.shape({
        groupId: PropTypes.number,
        members: PropTypes.arrayOf(PropTypes.shape({
            heroName: PropTypes.string,
            platformDisplayName: PropTypes.string,
            skillRating: PropTypes.number,
            stats: PropTypes.object,
        })),
        pending: PropTypes.array,
        leader: PropTypes.shape({
            heroName: PropTypes.string,
            platformDisplayName: PropTypes.string,
            skillRating: PropTypes.number,
            stats: PropTypes.object,
        }),
    }),
    preferredHeroes: PropTypes.shape({
        heroes: PropTypes.array
    }).isRequired,
    user: PropTypes.object.isRequired
};

const mapStateToProps = (state) => {
    return {
      group: state.group,
      preferredHeroes: state.preferredHeroes,
      user: state.user
    };
};

export default connect(mapStateToProps)(TimerGroupContainer);