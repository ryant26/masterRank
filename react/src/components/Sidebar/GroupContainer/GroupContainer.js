import React, {
    Component
} from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Model from '../../../model/model';
import MemberCard from './MemberCard/MemberCard';
import Modal from "../../Modal/Modal";
import GroupStatsContainer from "../../Stats/GroupStatsContainer";

class GroupContainer extends Component {

    constructor(props) {
        super(props);
        this.state = {
            showModal: false
        };

        this.toggleModal = this.toggleModal.bind(this);
    }

    componentWillUnmount() {
        Model.leaveGroup();
    }

    toggleModal() {
        this.setState({
            showModal: !this.state.showModal
        });
    }

    render() {
        let leader = this.props.group.leader;
        let members =  this.props.group.members;
        let pendingMembers = this.props.group.pending;

        return (
            <div className="GroupContainer">
                 <div>
                    { leader &&
                        <MemberCard
                            member={leader}
                            leader={true}
                            pending={false}
                            number={1}
                            key={[leader.platformDisplayName, leader.heroName]}
                        />
                    }
                    { members &&
                        members.map((member, i) => (
                            <MemberCard
                                member={member}
                                leader={false}
                                pending={false}
                                number={(i + 2)}
                                key={[member.platformDisplayName, member.heroName]}
                            />
                        ))
                    }
                    { pendingMembers &&
                        pendingMembers.map((member, i) => (
                            <MemberCard
                                member={member}
                                leader={false}
                                pending={true}
                                number={(members.length + 2 + i)}
                                key={member.platformDisplayName}
                            />
                        ))
                    }
                    <button className="button-four flex align-center justify-center" onClick={this.toggleModal}>
                        <div className="button-content">
                            Team Stats
                        </div>
                    </button>
                 </div>
                 <Modal modalOpen={this.state.showModal} closeModal={this.toggleModal}>
                     <GroupStatsContainer group={this.props.group} isLeading={true}/>
                 </Modal>
            </div>
        );
    }
}

GroupContainer.propTypes = {
    group: PropTypes.shape({
        groupId: PropTypes.number,
        members: PropTypes.arrayOf(PropTypes.shape({
            heroName: PropTypes.string,
            platformDisplayName: PropTypes.string,
            skillRating: PropTypes.number,
            stats: PropTypes.object,
        })),
        pending: PropTypes.array,
        leader: PropTypes.shape({
            heroName: PropTypes.string,
            platformDisplayName: PropTypes.string,
            skillRating: PropTypes.number,
            stats: PropTypes.object,
        }),
    }),
    user: PropTypes.object.isRequired
};

const mapStateToProps = (state) => {
    return {
      group: state.group,
      preferredHeroes: state.preferredHeroes,
      user: state.user
    };
};

export default connect(mapStateToProps)(GroupContainer);