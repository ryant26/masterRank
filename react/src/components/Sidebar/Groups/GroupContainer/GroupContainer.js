import React, {
    Component
  } from 'react';
import { connect } from 'react-redux';  
import PropTypes from 'prop-types';
import GroupHeroCard from '../GroupHeroCard/GroupHeroCard';
import HEROES from '../../../../resources/heroes';

export class GroupContainer extends Component {

    constructor(props) {
        super(props);
        this.state = { showing: false };
        this.createGroup = this.createGroup.bind(this);   
        this.leaveGroup = this.leaveGroup.bind(this);
        this.groupHeroes = this.props.groupHeroes;
        this.user = this.props.user;
    }

    createGroup() {
        // TODO: Create Group socket function
        // use 'create group' button -> fire serverEvents.createGroup
        // removes 'create group' button -> adds 'You' as initial card
        // use createGroupData -> socket server
        // creategroupdata === playerInvited from socket api wiki
    }

    leaveGroup() {
        // todo: Actually Implement
    }

    // Event Listener todo:
        // listen for group invite accepted event -> add group card to group list
        // listen for groupInviteDecline
        // listen for groupInviteCancel
        // listen for groupInviteSend from InvitePlayerButton to create and show pending text when user has not accepted invite
        // listen for playerInvited -> 
            // populate current the invited user's card with 'pending'
            // include a loading animation there: like ... dancing or something
            // timeout event after 30000 ms
               
    render() {
        // todo: implement get current users preferred hero from socket
        const usersPreferredHero = HEROES[0];
        const groupHeroCards = this.groupHeroes.map((hero,i) => {
            i = i + 2;
            return <GroupHeroCard hero={hero} name={hero.platformDisplayName} number={i.toString()} key={i.toString()} />;
        });

        return (
            <div className="GroupContainer sidebar-card flex flex-column">
            <div className="sidebar-title">Your Group</div>
                { this.state.showing 
                    ?   <div>  
                            <GroupHeroCard hero={usersPreferredHero} name={'You'} number={'1'} key={'1'} />
                            {groupHeroCards}
                            <button className="button-primary flex align-center justify-center">
                                <div className="button-content">
                                    Team Stats
                                </div>
                            </button>
                        </div>
                    :   <button className="button-primary" onClick={
                                (prevState) => {
                                    this.createGroup;
                                    this.setState({ showing: !prevState.showing });
                                }
                            }>
                            <div className="button-content">
                                Create Group
                            </div>
                        </button> 
                }
                <i className="leaveText" onClick={this.leaveGroup}>leave group</i>
            </div>
        );
    }
}

GroupContainer.propTypes = {
    user: PropTypes.object,
    groupHeroes: PropTypes.array.isRequired
};

const mapStateToProps = (state) => (
    {
      groupHeroes: state.groupHeroes
    }
);

export default connect(mapStateToProps)(GroupContainer);