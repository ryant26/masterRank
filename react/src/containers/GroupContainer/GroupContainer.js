import React, {
    Component
  } from 'react';
import { connect } from 'react-redux';  
import PropTypes from 'prop-types';
import GroupHeroCard from '../../components/GroupHeroCard/GroupHeroCard';

export class GroupContainer extends Component {

    constructor(props) {
        super(props);
        this.state = { showing: false };
        this.createGroup = this.createGroup.bind(this);   
        this.leaveGroup = this.leaveGroup.bind(this);
    }

    createGroup() {
        // TODO: Create Group socket function
        // use 'create group' button -> fire serverEvents.createGroup
        // removes 'create group' button -> adds 'You' as initial card
        // use createGroupData -> socket server
        // creategroupdata === playerInvited from socket api wiki
    }

    leaveGroup() {
        // TODO: Actually Implement
    }

    // Event Listener TODOs:
        // listen for group invite accepted event -> add group card to group list
        // listen for groupInviteDecline
        // listen for groupInviteCancel
        // listen for groupInviteSend from InvitePlayerButton to create and show pending text when user has not accepted invite
        // listen for playerInvited -> 
            // populate current the invited user's card with 'pending'
            // include a loading animation there: like ... dancing or something
            // timeout event after 30000 ms
               
    render() {
        const { group = [] } = this.props;
        let i = 0;

        const groupHeroCards = group.map(hero => {
            i = i + 1;
            return <GroupHeroCard hero={hero} number={i.toString()} key={i.toString()} />;
        });

        return (
            <div className="GroupContainer">
                <div className="inLine0">
                    <div className="sidebar-title">Your Group</div>
                </div>
                { this.state.showing 
                    ?   <div>  
                            {groupHeroCards}
                            <button className="button-primary">
                                <div className="button-content">
                                    Team Stats
                                </div>
                            </button>
                        </div>
                    :   <button className="button-primary" onClick={
                                (prevState) => {
                                    this.createGroup;
                                    this.setState({ showing: !prevState.showing });
                                }
                            }>
                            <div className="button-content">
                                Create Group
                            </div>
                        </button> 
                }
                <a className="leaveText" onClick={this.leaveGroup}><i>leave group</i></a>
            </div>
        );
    }
}

GroupContainer.propTypes = {
    user: PropTypes.object.isRequired,
    group: PropTypes.array.isRequired
};

const mapStateToProps = (state) => (
    {
      group: state.group
    }
);

export default connect(mapStateToProps)(GroupContainer);