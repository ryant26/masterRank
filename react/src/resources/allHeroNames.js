export const names = [
  'ana',
  'bastion',
  'dva',
  'genji',
  'hanzo',
  'junkrat',
  'lucio',
  'mccree',
  'mei',
  'mercy',
  'orisa',
  'pharah',
  'reaper',
  'reinhardt',
  'roadhog',
  'soldier76',
  'sombra',
  'symmetra',
  'torbjorn',
  'tracer',
  'widowmaker',
  'winston',
  'zarya',
  'zenyatta'
];
