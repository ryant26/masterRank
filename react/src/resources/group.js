const GROUP = [
    {
      "_id": 12,
      "platformDisplayName": "PwNShoPP#1662",
      "platform": "pc",
      "region": "us",
      "lastModified": "2017-10-28T06:08:42.474Z",
      "heroName": "soldier76",
      "hoursPlayed": 4,
      "wins": 10,
      "losses": 1,
      "kdRatio": 3.26,
      "accuracy": 43,
      "blockedPerMin": 0,
      "healingPerMin": 0,
      "damagePerMin": 1305.59,
      "avgObjElims": 9.16,
      "avgObjTime": 53.32,
      "pKdRatio": 0.75,
      "pAccuracy": 0.60,
      "pBlockedPerMin": 0,
      "pHealingPerMin": 0.25,
      "pDamagePerMin": 0.9,
      "pAvgObjElims": 0.5,
      "pAvgObjTime": 0.65,
    }, 
    {
      "_id": 12,
      "platformDisplayName": "santa#1662",
      "platform": "pc",
      "region": "us",
      "lastModified": "2017-10-28T06:08:42.474Z",
      "heroName": "dva",
      "wins": 10,
      "losses": 1,
      "kdRatio": 3.26,
      "hoursPlayed": 4,
      "accuracy": 43,
      "blockedPerMin": 0,
      "healingPerMin": 0,
      "damagePerMin": 1305.59,
      "avgObjElims": 9.16,
      "avgObjTime": 53.32,
      "pKdRatio": 0.75,
      "pAccuracy": 0.60,
      "pBlockedPerMin": 0,
      "pHealingPerMin": 0.25,
      "pDamagePerMin": 0.9,
      "pAvgObjElims": 0.5,
      "pAvgObjTime": 0.65,  },
    {
      "_id": 12,
      "platformDisplayName": "rustyK#1662",
      "platform": "pc",
      "region": "us",
      "lastModified": "2017-10-28T06:08:42.474Z",
      "heroName": "hanzo",
      "wins": 10,
      "losses": 1,
      "kdRatio": 3.26,
      "hoursPlayed": 4,
      "accuracy": 43,
      "blockedPerMin": 0,
      "healingPerMin": 0,
      "damagePerMin": 1305.59,
      "avgObjElims": 9.16,
      "avgObjTime": 53.32,
      "pKdRatio": 0.75,
      "pAccuracy": 0.60,
      "pBlockedPerMin": 0,
      "pHealingPerMin": 0.25,
      "pDamagePerMin": 0.9,
      "pAvgObjElims": 0.5,
      "pAvgObjTime": 0.65,  },
    {
      "_id": 12,
      "platformDisplayName": "wisesm0#1147",
      "platform": "pc",
      "region": "us",
      "lastModified": "2017-10-28T06:08:42.474Z",
      "heroName": "zenyatta",
      "wins": 10,
      "losses": 1,
      "kdRatio": 3.26,
      "hoursPlayed": 4,
      "accuracy": 43,
      "blockedPerMin": 0,
      "healingPerMin": 0,
      "damagePerMin": 1305.59,
      "avgObjElims": 9.16,
      "avgObjTime": 53.32,
      "pKdRatio": 0.75,
      "pAccuracy": 0.60,
      "pBlockedPerMin": 0,
      "pHealingPerMin": 0.25,
      "pDamagePerMin": 0.9,
      "pAvgObjElims": 0.5,
      "pAvgObjTime": 0.65,  },
  ];
  
  export default GROUP;
  