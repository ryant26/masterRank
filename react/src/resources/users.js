export const users = [
    {
        "_id": 12,
        "platformDisplayName": "PwNShoPP#1662",
        "lastUpdated": "2017-10-15T01:43:36.459Z",
        "platform": "pc",
        "level": 155,
        "portrait": "https://d1u1mce87gyfbn.cloudfront.net/game/unlocks/0x025000000000115A.png",
        "region": "us"
    }
];