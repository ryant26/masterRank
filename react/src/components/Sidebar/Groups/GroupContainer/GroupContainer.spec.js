import React from 'react';
import { mount } from 'enzyme';
import GROUP from '../../../../resources/group';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

import GroupContainer from './GroupContainer';
import * as users from '../../../../resources/users';

const mockStore = configureStore();

describe('Group Hero Container',()=> {
    let store;

    beforeEach(() => {
        store = mockStore({groupHeroes: GROUP});
    });
    
    it ('should render without exploding', () => {
        const wrapper = mount(
            <GroupContainer user={users.users[0]} store={store} />
        );
        
        const GroupContainerComponent = wrapper.find(GroupContainer);
        expect(GroupContainerComponent).toBeTruthy();
    });

    it ('should render a group leader as first member which is you', () => {
        const component = renderer.create(
            <GroupContainer store={store}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    xit ('should render two group joiners plus you', () => {
    
    });

    xit ('should render three group joiners plus you', () => {
        
    });

    xit ('should render four group joiners plus you', () => {
        
    });

    xit ('should render five group joiners plus you', () => {
        
    });
});
