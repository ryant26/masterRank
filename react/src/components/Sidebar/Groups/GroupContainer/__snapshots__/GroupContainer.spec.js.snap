// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group Hero Container should render a group leader as first member which is you 1`] = `
<div
  className="GroupContainer sidebar-card flex flex-column"
>
  <div
    className="sidebar-title"
  >
    Your Group
  </div>
  <button
    className="button-primary"
    onClick={[Function]}
  >
    <div
      className="button-content"
    >
      Create Group
    </div>
  </button>
  <i
    className="leaveText"
    onClick={[Function]}
  >
    leave group
  </i>
</div>
`;
